flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java

```diff
@@ -333,28 +336,28 @@ private static void restartTaskManager(Runnable afterFailAction, MiniCluster min
   private static class RecordCounterToFail {
 
     private static AtomicInteger records;
-    private static CompletableFuture<Void> fail;
+    private static CountDownLatch countDownLatch;
     private static CompletableFuture<Void> continueProcessing;
 
     private static <T> DataStream<T> wrapWithFailureAfter(DataStream<T> stream, int failAfter) {
 
       records = new AtomicInteger();
-      fail = new CompletableFuture<>();
       continueProcessing = new CompletableFuture<>();
+      countDownLatch = new CountDownLatch(stream.getParallelism());
       return stream.map(
           record -> {
             boolean reachedFailPoint = records.incrementAndGet() > failAfter;
-            boolean notFailedYet = !fail.isDone();
+            boolean notFailedYet = countDownLatch.getCount() != 0;
             if (notFailedYet && reachedFailPoint) {
-              fail.complete(null);
+              countDownLatch.countDown();
               continueProcessing.get();
             }
             return record;
           });
     }
 
-    private static void waitToFail() throws ExecutionException, InterruptedException {
-      fail.get();
+    private static void waitToFail() throws InterruptedException {
```

Comment:
I also just realized that this method is probably not accurate anymore with the `testBoundedWithSavepoint`. Maybe it should be called `waitForCondition`? also the inner class name `RecordCounterToFail` could be renamed to `RecordCounterToWait`.

Reply:
this will also invalidate the class name `TestIcebergSourceFailover`. It was added here probably for code reuse. otherwise, the better place for `testBoundedWithSavepoint ` is probably `TestIcebergSourceFailover`

Reply:
I don't get this last comment 😢

Reply:
sorry for the typo. I meant `TestIcebergSourceBounded` is probably a better home for `testBoundedWithSavepoint` if we can figure out the code reuse from `TestIcebergSourceFailover`. Otherwise, we need to rename `TestIcebergSourceFailover` to accurately reflect that it doesn't just test failover. It also tests savepoint trigger and no duplication.

Reply:
The issue was discovered when there was a job restart, because of a temporary issue in a downstream service.
This is a failure recovery scenario, in many ways very similar to the Failover cases tested here. Maybe we can rename the class to `TestIcebergSourceFailureRecovery`?

That said, I see the following issues mixing refactoring and actual fixes in the same PR:
- Makes it hard in retrospect to understand what was the fix, and what was the refactor. If we need to revisit the fix, because it's not working, or we have a similar issue, then we will have hard time separating those. 
- If you have a stable branch for internal releases, you might not want to accept the risks inherent to the refactor, but you might need the fix. Mixing fix and refactor code makes this hard.
- If you refactor the code that you are fixing, and move it to another class, then it is very hard to spot small mistakes or changes done in the same part of the code. Makes the reviewer life hard, and the process error prone. 
- Blocking flaky test fix on refactoring could waste serious developer hours, if other developers bumps into the same issue and sink time into trying to find and fix the bug.

I am all for cleaning up code, and happy to do it whenever I have time, but I think that if it's not related to the actual fix/PR, then it should be done only with a very limited scope. If we spot something which needs to be fixed, we should create a specific, refactoring PR for it. I would be happy to review it, whenever you make one.

Reply:
I didn't mean we have to do it in this PR. it is definitely reasonable to separate it out.

while we are at it, I just noticed the intention and coverage of this class has changed. hence want to point it out.